server/service/microsoft_mdm.go

```diff
@@ -1248,17 +1249,125 @@ func (svc *Service) isTrustedRequest(ctx context.Context, reqSyncML *fleet.SyncM
 	return errors.New("calling device is not trusted")
 }
 
+// isValidUPN checks if the provided user ID is a valid UPN
+func isValidUPN(userID string) bool {
+	const upnRegex = `^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`
+	re := regexp.MustCompile(upnRegex)
```

Comment:
Nit, but instead of compiling the regexp on each call of this function, it should be compiled just once at startup and reused afterwards, i.e. extracting those two lines outside the function and having a variable there:

```
var rxValidUPN = regexp.MustCompile(`^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`)
```